launchUI.py

```diff
@@ -1571,69 +1567,77 @@ def place_xbar_options(master):
     configMemEntry = ttk.Entry(paramPannel, justify=tkinter.CENTER)
-    configMemEntry.grid(row=1, column=3, sticky=tkinter.W)
+    configMemEntry.grid(row=1, column=2)
     configMemEntry.insert(0, paramCGRA.configMemSize)
     widgets["configMemEntry"] = configMemEntry
     
     dataMemLabel = ttk.Label(paramPannel, text='Data SPM (KBs):')
-    dataMemLabel.grid(columnspan=2, row=2, column=0,sticky=tkinter.W)
+    dataMemLabel.grid(row=2, column=0, sticky=tkinter.W)
     dataMemEntry = ttk.Entry(paramPannel, justify=tkinter.CENTER)
-    dataMemEntry.grid(row=2, column=2, sticky=tkinter.W)
+    dataMemEntry.grid(row=2, column=1)
     dataMemEntry.insert(0, str(paramCGRA.dataMemSize))
     widgets["dataMemEntry"] = dataMemEntry   
     updateButton = tkinter.Button(paramPannel, text = " Reset ", relief='raised', command = partial(clickReset, master))
-    updateButton.grid(row=2, column=3)
+    updateButton.grid(row=2, column=2)
 
     entireTileCheckVar.set(0)
     entireTileCheckbutton = tkinter.Checkbutton(paramPannel, variable=entireTileCheckVar, text="Disable entire Tile 0", command=clickEntireTileCheckbutton)
-    entireTileCheckbutton.grid(columnspan=3, row=3, column=0,sticky="W")
+    entireTileCheckbutton.grid(columnspan=2, row=3, column=0, sticky="W")
     widgets["entireTileCheckbutton"] = entireTileCheckbutton
     resetButton = tkinter.Button(paramPannel, text = "Update", relief='raised', command = partial(clickUpdate, master))
-    resetButton.grid(row=3, column=3)
+    resetButton.grid(row=3, column=2)
 
     fuConfigPannel = tkinter.LabelFrame(paramPannel, text='Tile 0 functional units', bd = BORDER, relief='groove')
-    fuConfigPannel.grid(columnspan=4, row=4, column=0,rowspan=3,sticky="nsew")
-    widgets["fuConfigPannel"] = fuConfigPannel    
+    fuConfigPannel.grid(columnspan=4, row=4, column=0, rowspan=3, sticky="nsew")
+    widgets["fuConfigPannel"] = fuConfigPannel
+    fuConfigPannel.columnconfigure(0, weight=1)
+    fuConfigPannel.columnconfigure(1, weight=1)
+    fuConfigPannel.columnconfigure(2, weight=1)
+    fuConfigPannel.columnconfigure(3, weight=1) # Use columnconfigure and rowconfigure to partition the columns, so that each sub-module fills the corresponding space  
     place_fu_options(fuConfigPannel)
 
     xbarConfigPannel = tkinter.LabelFrame(paramPannel, text='Tile 0 crossbar outgoing links', bd=BORDER, relief='groove')
-    xbarConfigPannel.grid(columnspan=4, row=7, column=0,rowspan=2,sticky="nsew")
+    xbarConfigPannel.grid(columnspan=4, row=7, column=0, rowspan=2, sticky="nsew")
     widgets["xbarConfigPannel"] = xbarConfigPannel
+    xbarConfigPannel.columnconfigure(0, weight=1)
+    xbarConfigPannel.columnconfigure(1, weight=1)
+    xbarConfigPannel.columnconfigure(2, weight=1)
+    xbarConfigPannel.columnconfigure(3, weight=1) # Use columnconfigure and rowconfigure to partition the columns, so that each sub-module fills the corresponding space  

```

Comment:
Where is `rowconfigure`? So there are 4 columns? What are the 4 sub-modules?

Reply:
I think I used an inappropriate description. Using "sub-modules" to describe it is not very appropriate. I arranged the 12 function buttons in the xbar window in 3 rows and 4 columns. The code part is to align the 4 divided columns in the xbarpannel so that they can change with the window size.